client-spark/spark3/src/main/java/org/apache/spark/shuffle/MemoryLimitedMap.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Tencent is pleased to support the open source community by making
+ * Firestorm-Spark remote shuffle server available.
+ *
+ * Copyright (C) 2021 THL A29 Limited, a Tencent company.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package org.apache.spark.shuffle;
+
+import org.apache.spark.memory.TaskMemoryManager;
+import org.apache.spark.util.collection.SizeTrackingAppendOnlyMap;
+import org.apache.spark.util.collection.Spillable;
+import scala.Function1;
+import scala.Function2;
+import scala.Tuple2;
+import scala.collection.Iterator;
+
+/**
+ * This class is the wrapper of {@link SizeTrackingAppendOnlyMap} and to limit the
+ * memory usage from the task memory manager. Once the memory usage exceed the
+ * {@code spillThresholdSize} or lack the enough memory from task memory manager,
+ * it will spill elements by the external function of {@code spillFunc}.
+ */
+public class MemoryLimitedMap<K, V> extends Spillable<SizeTrackingAppendOnlyMap> {
+    private SizeTrackingAppendOnlyMap<K, V> currentMap = new SizeTrackingAppendOnlyMap<>();
+    private final long spillThresholdSize;
+    private final Function1<Tuple2<K, V>, Void> spillFunc;
+
+    public MemoryLimitedMap(TaskMemoryManager taskMemoryManager, Long spillThresholdSize,
+            Function1<Tuple2<K, V>, Void> spillFunc) {
+        super(taskMemoryManager);
+        this.spillThresholdSize = spillThresholdSize;
+        this.spillFunc = spillFunc;
+    }
+
+    @Override
+    public void spill(SizeTrackingAppendOnlyMap map) {
+        Iterator<Tuple2<K, V>> iter = map.iterator();
+        while (iter.hasNext()) {
```

Comment:
It looks unnecessary, refer to https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/util/collection/Spillable.scala#L82

It will be handled in the method of maybeSpill

Reply:
So we shouldn't call method `spill` directly. We should call method `spill` through `maySpill`.

Reply:
Yep.  Agree with u.

However, the abstract class of `spillable` will not check the memory usage exceed our specified 'rss.client.mapside.merge.spill.threshold'. If we override the method of `maybeSpill`, some vars is limited to access due to access modifier.

So i will make the memoryLimitedMap extends the `MemoryConsumer` directly instead of `Spillable` ?  Do u think so?

Reply:
> Yep. Agree with u.
> 
> However, the abstract class of `spillable` will not check the memory usage exceed our specified 'rss.client.mapside.merge.spill.threshold'. If we override the method of `maybeSpill`, some vars is limited to access due to access modifier.
> 
> So i will make the memoryLimitedMap extends the `MemoryConsumer` directly instead of `Spillable` ? Do u think so?

Could you use refection to solve the access modifier problems?

Reply:
I think reflection is ugly and wont fail on compiling when upstream codebase changed if the method changed.

Reply:
Now we only adapt the fixed release version's Spark,  the upstream codebase won't change actually.

Reply:
Maybe we should use test case to find changes of method. But if the work is not too much,  it's ok to make the MemoryLimitedMap extends the `MemoryConsumer`.